includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -620,6 +645,11 @@ public function process_payment_for_order( $order, $cart, $payment_information,
 	protected function add_token_to_order( $order, $token ) {
 		$order_tokens = $order->get_payment_tokens();
 
+		// This could lead to tokens being saved twice in an order's payment tokens, but it is needed so that shoppers
+		// may re-use a previous card for the same subscription, as we consider the last token to be the active one.
+		// We can't remove the previous entry for the token because WC_Order does not support removal of tokens [1] and
```

Comment:
The list of tokens associated with an order is just a simple array of IDs in the order metadata ([source](https://github.com/woocommerce/woocommerce/blob/d48f1d4e2efcfab8206799df2085f021ec273722/includes/data-stores/abstract-wc-order-data-store-cpt.php#L426)). Now, what is worse? Manipulating the raw metadata, or keep relying on the last token of an order being the only one that "counts"? There's no easy answer :)

Reply:
Good question. I feel like relying on the last token might cause some issues if WC's underlying code changes and the order is not preserved, similarly to what happened with https://github.com/Automattic/woocommerce-payments/pull/832/commits/cfc0619cf6ac94d4bf728e3b48f325bd3077d443, and handling the raw metadata for that public property could be a cause for some conflicts with other extensions that might overwrite it.

Having another metadata for the payment method would make it easier to manage it, but on the other hand, it "pollutes" the order metadata as it's not [considered internal like the tokens]( https://github.com/woocommerce/woocommerce/blob/d48f1d4e2efcfab8206799df2085f021ec273722/includes/data-stores/abstract-wc-order-data-store-cpt.php#L44).

cc @reykjalin, @RadoslavGeorgiev - what do you guys think?

Reply:
I think relying on the WC core behaviour here is probably the way to go. The less we manually manipulate data in the database the better imo 🙂.

I especially don't think we should mess directly with the metadata field that's in use here. **If** we decide to store and manipulate something in the database directly for more control I think we should store it somewhere that's used only for wcpay.

Messing with the metadata directly here doesn't sit right with me 🙂